Feat free properties as filter zweite welle

laurgarn opened the pull request.

laurgarn committed:
fix: add some space between name and values

laurgarn committed:
refact:code split

laurgarn committed:
not so bad in read mode

laurgarn committed:
feat: add sectioning + fix one side filter css + remove some console

laurgarn committed:
feat: roundtrip OK

laurgarn committed:
feat: red only shrinks between validations + PropertyValues1 and PropertyValues2

dbraquart commented:
"zweite welle" in PR title ??

laurgarn committed:
fixrev:equipment type for HVDC is HVDC_LINE

laurgarn committed:
Merge remote-tracking branch 'origin/main' into feat_free_properties_as_filter_02

dbraquart commented on src/utils/equipment-types.js:
```diff
@@ -87,6 +75,19 @@ const FreeProps = {
         type: filteredTypes.freeProperties,
     },
 };
+const FreePropsS = {
```
S for substation ? you can be more explicit in naming

laurgarn replied:
We are happy with '1' and '2'. Why could we not be happy with just this 'S' that has just one local usage ?

dbraquart replied:
hum.. not so local, you have a corresponding component named FreePropertiesS

dbraquart commented on src/components/dialogs/create-filter-dialog.js:
```diff
@@ -112,14 +112,16 @@ const CreateFilterDialog = ({
     const [id, setId] = useState('');
-    const handleCallback = (criteriaFilter) => {
-        if (criteriaFilter) {
+    const handleCallback = (criteriaFilter, veto) => {
```
what this flag "veto" really means ?

laurgarn replied:
"Veto" is usually used to tell "I/we don't want", see https://en.wiktionary.org/wiki/veto. And this is what it means here, as it is use also in https://docs.oracle.com/javase/7/docs/api/index.html?java/beans/VetoableChangeListener.html
The method in itself is not that long and it can quite easily be infered that if veto is true, no filter is to be saved (and even be reset).

dbraquart commented on src/components/dialogs/create-filter-dialog.js:
```diff
@@ -209,8 +211,9 @@ const CreateFilterDialog = ({
 
     const handleValidation = () => {
-        //To manage the case when we never tried to enter a name
+        setValidationsCount((prev) => prev + 1);
```
I dont understand the purpose of this counter ? maybe we can discuss about it, dont reply here.

laurgarn replied:
It carries the information that validation number has changed or not, allowing to not mark as a problem something like an empty property line after it has been added but has it marked after a new validation.

dbraquart commented on src/components/dialogs/criteria-based-filter-dialog-content.js:
```diff
@@ -83,12 +84,80 @@ export const FilterTypeSelection = ({
     );
 };
 
+const backToFrontTweak = (response) => {
```
maybe you could comment these 2 functions. They are very "technical", it's hard to follow what they really do

laurgarn replied:
It just adapts the data structure from what is received from the backend to the form that the component can handle.

dbraquart commented on src/components/dialogs/criteria-based-filter-dialog-content.js:
```diff
@@ -115,14 +184,12 @@ export const CriteriaBasedFilterDialogContent = ({
-                            response.equipmentFilterForm.equipmentType
-                        );
+                        setInitialFilter(backToFrontTweak(response));
+                        let eType = response.equipmentFilterForm.equipmentType;
```
etype could be const

laurgarn replied:
Yes, it could. But it would trigger the 80 characters per line limit that bother me more.

dbraquart commented on src/components/dialogs/expandable-criteria.js:
```diff
@@ -0,0 +1,134 @@
+    onChange,
+    validateItems,
+}) => {
+    // const classes = useStyles();
```
remove it

laurgarn replied:
Well spotted

dbraquart commented on src/components/dialogs/expandable-criteria.js:
```diff
@@ -0,0 +1,134 @@
+        values,
+        fieldProps,
+        // classes.button,
+        // classes.deleteButton,
```
remove comments

laurgarn replied:
Thx

dbraquart commented on src/components/dialogs/free-properties-based-filter.js:
```diff
@@ -0,0 +1,517 @@
+                ? {}
+                : Object.fromEntries(arr.map((p) => [p.name || '', p.values]));
+            const vetoIdx = arr.findIndex((p) => !p.name || !p?.values?.length);
+            onChange(obj, vetoIdx >= 0);
```
here is veto again ... a comment could help. What are we doing in this function ?

dbraquart reviewed (changes requested):
review: a few comments
tests: 2 ko
a) unable to add/create a filtering by property (+ button), for eqpt "HVDC Lines".
The sent Json sounds bad, it does not use freeProperties1 / freeProperties2 like with the Lines.

b)
1- create Filter on line (or hvdc line), with at least one properties line :region/east/west + validate
2- edit it, switch to another eqpt => crash
 CardErrorBoundary caught:  TypeError: value is undefined
    useAutocomplete useAutocomplete.js:916
    Autocomplete Autocomplete.js:423

remark: when I enter twice the same properties name, I still can validate the Form (the second properties line is deleted)

Console warning: I got 3 warnings:
1) Warning: Failed prop type: Invalid prop `columns[0].minWidth` of type `string` supplied to `MuiVirtualizedTable`, expected `number`.
2)  Warning: Failed prop type: Invalid prop `minWidth` of type `string` supplied to `Column`, expected `number`.
3)  Warning: Cannot update a component (`CriteriaBasedFilterDialog`) while rendering a different component (`SingleFilter`).
 To locate the bad setState() call inside `SingleFilter`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

laurgarn commented:
> review: a few comments tests: 2 ko
> a) unable to add/create a filtering by property (+ button), for eqpt "HVDC Lines". The sent Json sounds bad, it does not use freeProperties1 / freeProperties2 like with the Lines.

addressed and corrected.

> b) 1- create Filter on line (or hvdc line), with at least one properties line :region/east/west + validate
>     2- edit it, switch to another eqpt => crash CardErrorBoundary caught: TypeError: value is undefined useAutocomplete useAutocomplete.js:916 Autocomplete Autocomplete.js:423

I'll address this. To trigger this, we need to switch to a none line or hvdc

> remark: when I enter twice the same properties name, I still can validate the Form (the second properties line is deleted)
> 
> Console warning: I got 3 warnings:
> 
>     1. Warning: Failed prop type: Invalid prop `columns[0].minWidth` of type `string` supplied to `MuiVirtualizedTable`, expected `number`.
> 
>     2. Warning: Failed prop type: Invalid prop `minWidth` of type `string` supplied to `Column`, expected `number`.

These first two have nothing to do with this PR.
They have with me, because I told Walid to use a css compatible size as column minWidth, forgetting our MUIVirtualizedTable sizes method does not handle that smartly enough.
But this has nothing to do we the changed code of this PR.

>     3. Warning: Cannot update a component (`CriteriaBasedFilterDialog`) while rendering a different component (`SingleFilter`).
>        To locate the bad setState() call inside `SingleFilter`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

laurgarn committed:
fixrev:crash on eqpt type change

laurgarn committed:
fixrev:same name case prevented OK + ExpandableCriteria as component

laurgarn committed:
Merge remote-tracking branch 'origin/main' into feat_free_properties_as_filter_02

laurgarn committed:
doc:add dome jsdoc to FreeProperties, FreeProperties2, FreeProperty & FreeProperty2

 + remove export to FreeProperty & FreeProperty2

laurgarn committed:
fix:substation validationsCount was not transmitted.

laurgarn committed:
fix:global veto and equipment type change

laurgarn committed:
doc:some more comments

dbraquart commented on src/components/dialogs/free-properties-based-filter.js:
```diff
@@ -0,0 +1,577 @@
+    );
+};
+
+const makeFreePropertiesProblemsMap = (values, full, prevErrors) => {
```
this function is a clone of makeFreeProperties2ProblemsMap, except one line.
could we use a single function ? (proposal)

dbraquart replied:
review: ok (1 remark below)
console: ok
tests : ok

laurgarn committed:
fixrev: facorize out makeFreePropertiesProblemsMap and makeFreeProperties2ProblemsMap

dbraquart reviewed (approved):
review: ok
console: ok
tests : ok
